server/src/main/java/org/elasticsearch/index/mapper/MapperMergeContext.java

```diff
@@ -43,10 +49,113 @@ private MapperMergeContext(MapperBuilderContext mapperBuilderContext) {
      * @return a new {@link MapperMergeContext} with this context as its parent
      */
     public MapperMergeContext createChildContext(String name) {
-        return new MapperMergeContext(mapperBuilderContext.createChildContext(name));
+        return createChildContext(mapperBuilderContext.createChildContext(name));
+    }
+
+    /**
+     * Creates a new {@link MapperMergeContext} with a given {@link MapperBuilderContext}
+     * @param childContext the child {@link MapperBuilderContext}
+     * @return a new {@link MapperMergeContext}, wrapping the provided {@link MapperBuilderContext}
+     */
+    public MapperMergeContext createChildContext(MapperBuilderContext childContext) {
+        return new MapperMergeContext(childContext, remainingFieldsBudget);
     }
 
     MapperBuilderContext getMapperBuilderContext() {
         return mapperBuilderContext;
     }
+
+    void removeRuntimeField(Map<String, RuntimeField> runtimeFields, String name) {
+        if (runtimeFields.containsKey(name)) {
+            runtimeFields.remove(name);
+            remainingFieldsBudget.increment(1);
+        }
+    }
+
+    <M extends Mapper> boolean addFieldIfPossible(M mapper, Consumer<M> addField) {
+        if (remainingFieldsBudget.decrementIfPossible(mapper.mapperSize())) {
+            addField.accept(mapper);
+            return true;
+        }
+        return false;
+    }
+
+    void addRuntimeFieldIfPossible(RuntimeField runtimeField, Consumer<RuntimeField> addField) {
+        if (remainingFieldsBudget.decrementIfPossible(1)) {
+            addField.accept(runtimeField);
+        }
+    }
+
+    boolean hasRemainingBudget() {
+        return remainingFieldsBudget.hasRemainingBudget();
+    }
+
+    /**
+     * Keeps track of now many new fields can be added during mapper merge
+     */
+    public interface NewFieldsBudget {
```

Comment:
I've made it a private interface now. If we wanted to keep it public and pass it around instead of a long, we could also make it a sealed interface that only permits the `Unlimited` and `Limited` implementations.